tests/Unit/Core/Search/Builder/PersistFiltersBuilderTest.php

```diff
@@ -0,0 +1,275 @@
+        $this->assertNotNull($filters);
+        $this->assertEquals($expectedFilters, $filters->all());
+        $this->assertEquals('language', $filters->getFilterId());
+    }
+
+    public function testBuildWithRequest()
+    {
+        $expectedFilters = [
+            'limit' => 10,
+            'offset' => 10,
+        ];
+        $inputFilters = new Filters($expectedFilters);
+        $savedFilters = [
+            'limit' => 10,
+        ];
+
+        $builder = new PersistFiltersBuilder(
+            $this->buildRepositoryByRouteMock($savedFilters, 'language', 'index'),
+            $this->buildEmployeeProviderMock(),
+            static::SHOP_ID
+        );
+        $builder->setConfig([
+            'request' => $this->buildRequestMock('PrestaShopBundle\Controller\Admin\Improve\International\LanguageController::indexAction'),
+        ]);
+
+        $filters = $builder->buildFilters($inputFilters);
+        $this->assertNotNull($filters);
+        $this->assertEquals($expectedFilters, $filters->all());
+        $this->assertEmpty($filters->getFilterId());
+    }
+
+    public function testNoNeedForPersist()
+    {
+        $expectedFilters = [
+            'limit' => 10,
+            'offset' => 10,
+        ];
+        $inputFilters = new Filters($expectedFilters, 'language');
+        $inputFilters->setNeedsToBePersisted(false);
+
+        $builder = new PersistFiltersBuilder(
+            $this->buildUnusedRepository(),
+            $this->buildEmployeeProviderMock(1),
+            static::SHOP_ID
+        );
+
+        $filters = $builder->buildFilters($inputFilters);
+        $this->assertNotNull($filters);
+        $this->assertEquals($expectedFilters, $filters->all());
+        $this->assertEquals('language', $filters->getFilterId());
+    }
+
+    /**
+     * @param array $filters
+     * @param string $filterId
+     *
+     * @return MockObject|AdminFilterRepository
+     */
+    private function buildRepositoryByFilterIdMock(array $filters, $filterId)
+    {
+        $repositoryMock = $this->getMockBuilder(AdminFilterRepository::class)
+            ->disableOriginalConstructor()
+            ->getMock()
+        ;
+
+        $repositoryMock
+            ->expects($this->once())
+            ->method('createOrUpdateByEmployeeAndFilterId')
+            ->with(
+                $this->equalTo(static::EMPLOYEE_ID),
+                $this->equalTo(static::SHOP_ID),
+                $this->equalTo($filters),
+                $this->equalTo($filterId)
+            )
+        ;
+
+        $repositoryMock
+            ->expects($this->never())
+            ->method('createOrUpdateByEmployeeAndRouteParams')
+        ;
+
+        return $repositoryMock;
+    }
+
+    /**
+     * @param array $filters
+     * @param string $controller
+     * @param string $action
+     *
+     * @return MockObject|AdminFilterRepository
+     */
+    private function buildRepositoryByRouteMock(array $filters, $controller, $action)
```

Comment:
```suggestion
    private function buildRepositoryByRouteMock(array $filters, $controller, $action): AdminFilterRepository
```

Reply:
I thought it would scream because it returns a `MockObject` which wouldn't match the return type But surprisingly it does work without crashing!!

That said I prefer to keep the loose (absent) return type because setting it actually fires even more warning since the PHPDOc states that it can return `MockObject|AdminFilterRepository` Both are correct, one is the actual object, the other the mocked object (useful for auto completion and please the injected object constructor)

And if I use either one as a strict type then I get errorr/warning that it doesn't match the PHPDoc. So in this case, I think it's preferable and more accurate not to force strict type.

Reply:
I know it works. I already tested. :stuck_out_tongue_closed_eyes: 

I think it's preferable to strict types because it avoids a lot of work for PHPStan on tests directory. And the constructor will not like MockObject as parameter.